Start DM on first message (UI)

Florian14 opened the pull request:
## Type of change

- [X] WIP Feature
- [ ] Bugfix
- [ ] Technical
- [ ] Other :

## Content

This PR displays a "fake" local timeline after selecting the users to create a DM. Sending a message or event is not handled yet, so the real room cannot be created for now.
The flow is hidden behind a debug feature flag accessible from the debug menu.

## Motivation and context

UI: https://github.com/vector-im/element-android/issues/5525

## Screenshots / GIFs

https://user-images.githubusercontent.com/11990514/177182407-61af8ca8-e78e-4bf7-b01b-3bbe8aef492f.mp4

## Tests

- Activate the start DM feature in the debug menu -> Features
- On the home screen, select the first tab (Direct Messages)
- Click on the fab button to start a DM, select several users and press "GO"
- A kind of timeline should now appear, there is no action in the toolbar (except the close button)
- It is not possible to access the room settings or details
- Composer actions are not working (an error appears when we try to send an event to the room)
- When we go back to the room list or relaunch the app, the "fake" room is deleted from the DB and disappears from the room list

## Tested devices

- [ ] Physical
- [X] Emulator
- OS version(s): API 32

## Checklist

<!-- Depending on the Pull Request content, it can be acceptable if some of the following checkboxes stay unchecked. -->

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [X] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [ ] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [X] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [X] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

Florian14 commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/create/CreateRoomTask.kt:
```diff
@@ -44,9 +44,14 @@ import org.matrix.android.sdk.internal.util.awaitTransaction
 
 internal interface CreateRoomTask : Task<CreateRoomParams, String>
 
+@Qualifier
+@Retention(AnnotationRetention.BINARY)
+annotation class Default
+
```
I added a second implementation of the interface, so I added an annotation for each implementation to be able to choose which one I want to inject, is it OK?

ouchadam commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/create/CreateLocalRoomTask.kt:
```diff
@@ -0,0 +1,269 @@
+        private val createRoomBodyBuilder: CreateRoomBodyBuilder,
+        private val userService: UserService,
+        private val clock: Clock,
+) : CreateRoomTask {
```
do we have any usages of the base type? :eyes:  I'd be tempted to type this to a dedicated `DefaultLocalRoomTask` to match the `DefaultDeleteLocalRoom`

ouchadam commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/create/CreateLocalRoomTask.kt:
```diff
@@ -0,0 +1,269 @@
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+import javax.inject.Qualifier
+import org.matrix.android.sdk.internal.database.awaitNotEmptyResult as awaitNotEmptyResult1
```
does this need to be renamed? can't see a usage of `awaitNotEmptyResult`

ouchadam commented:
we can use the `.wip` extension for creating work in progress (behind a feature flag) change log entries

Florian14 commented on vector/src/main/res/layout/item_timeline_event_merged_room_creation_stub.xml:
```diff
@@ -65,28 +64,29 @@
             app:layout_constraintEnd_toEndOf="parent"
-            app:layout_constraintStart_toStartOf="parent"
-            app:layout_constraintTop_toBottomOf="@id/creationTileRoomAvatarImageView"
+            app:layout_constraintStart_toEndOf="@id/creationTileRoomAvatarImageView"
```
It does not work for long room names, the room name overlaps some views, I have to check how to fix it (changing some constraints or adding a max lines property)

Florian14 replied:
Fixed by moving the room name below the avatar

ouchadam commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/create/CreateLocalRoomTask.kt:
```diff
@@ -0,0 +1,267 @@
+                continue
+            }
+
+            eventIds.add(event.eventId)
```
it looks like these `eventIds` aren't used? :eyes:

Florian14 replied:
you're right, it's not used anymore 👍🏻

ouchadam commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/create/CreateLocalRoomTask.kt:
```diff
@@ -0,0 +1,267 @@
+
+        // Wait for room to be created in DB
+        try {
+            awaitNotEmptyResult(realmConfiguration, TimeUnit.MINUTES.toMillis(1L)) { realm ->
```
for my understanding, are we relying on another task to asynchronously update the room entity? if possible it would be safer to sequentially run the tasks rather than rely on a timeout (is this a common pattern in the sdk?)

Florian14 replied:
@ouchadam I agree with you, I'd like to remove it but I have a non-systematic but recurrent crash when I try to navigate to the local room (in the stacktrace, it seems that we try to navigate to un unknown room). It's a bit confusing because the awaitTransaction seems ok, I can log the entity before returning the roomId from the task. Did I miss something else?

ouchadam replied:
I'm not too familiar with this area, would need to delegate to @bmarty / @ganfra to answer that

it looks like other tasks are using the same pattern of `awaitNotEmptyResult`, not a blocker for me, was curious :smile:

Florian14 replied:
The second task is waiting to get the room from the sync but in the current case, the room is created locally and put in DB, so we should not have to wait for it 😕

bmarty replied:
Sorry for the cross conversations, I did not see @ouchadam 's comments at first sight.
Please see my suggestion at https://github.com/vector-im/element-android/pull/6051#discussion_r912992010

Florian14 replied:
> Maybe just replacing with this line can help (I let you do some test):
> ```kotlin
>         // Ensure that Realm is up to date before returning the roomId.
>         monarchy.doWithRealm { it.refresh() }
> ```

It continues crashing...

ouchadam commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/room/RoomService.kt:
```diff
@@ -40,6 +40,18 @@ interface RoomService {
+     * This room will not be synchronized with the server and will not come back from the sync, so all the events related to this room will be generated
+     * locally.
+     */
+    suspend fun createLocalRoom(createRoomParams: CreateRoomParams): String
```
thinking out loud about about the api design~ not a blocker!

from a consumer api perspective, are local rooms semantically the same as _normal_ rooms? / would we handle interacting with rooms differently if they're normal vs local?

I'm wondering if `CreateRoomParams` could have a `isLocal` option and reusing `createRoom` and `leaveRoom` :thinking:

ouchadam reviewed (approved):
LGTM :100:

fedrunov reviewed (approved):
LGTM

Florian14 commented:
@gaelledel I moved the room name below the avatar, updated some font styles and changed the description. See the following screenshot:
<img width="406" alt="Capture d’écran 2022-05-25 à 17 14 52" src="https://user-images.githubusercontent.com/11990514/170301425-cd3f56db-838e-4b31-b551-1e57773c9933.png">

If we look at the design, there are still some works to do (especially on the encryption tile and the horizontal margins). I prefer to do it in a future PR, as it will require more changes and may cause some bugs.

bmarty commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/create/CreateRoomTask.kt:
```diff
@@ -62,11 +62,6 @@ internal class DefaultCreateRoomTask @Inject constructor(
-        val otherUserId = if (params.isDirect()) {
-            params.getFirstInvitedUserId()
-                    ?: throw IllegalStateException("You can't create a direct room without an invitedUser")
-        } else null
```
Why this check has been removed?

bmarty commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/delete/DeleteLocalRoomTask.kt:
```diff
@@ -0,0 +1,77 @@
+                        ?.also { Timber.i("## DeleteLocalRoomTask - TimelineEventEntity - delete ${it.size} entries") }
+                        ?.deleteAllFromRealm()
+                ChunkEntity.where(realm, roomId = roomId).findAll()
+                        ?.also { Timber.i("## DeleteLocalRoomTask - RoomMemberSummaryEntity - delete ${it.size} entries") }
```
```suggestion
                        ?.also { Timber.i("## DeleteLocalRoomTask - ChunkEntity - delete ${it.size} entries") }
```

bmarty commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/delete/DeleteLocalRoomTask.kt:
```diff
@@ -0,0 +1,77 @@
+                TimelineEventEntity.whereRoomId(realm, roomId = roomId).findAll()
+                        ?.also { Timber.i("## DeleteLocalRoomTask - TimelineEventEntity - delete ${it.size} entries") }
+                        ?.deleteAllFromRealm()
+                ChunkEntity.where(realm, roomId = roomId).findAll()
```
Could also use `ChunkEntity.deleteOnCascade` which does deep cleaning for you.

bmarty commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/timeline/DefaultTimeline.kt:
```diff
@@ -389,6 +390,7 @@ internal class DefaultTimeline(
     }
 
     private suspend fun loadRoomMembersIfNeeded() {
+        if (RoomLocalEcho.isLocalEchoId(roomId)) return
```
I think this check could be avoided if we store the fact that all the members are known for local rooms.

bmarty commented on vector/src/debug/java/im/vector/app/features/debug/features/DebugVectorFeatures.kt:
```diff
@@ -116,11 +119,12 @@ private fun <T : Enum<T>> enumPreferencesKey(type: KClass<T>) = stringPreference
     val onboardingAlreadyHaveAnAccount = booleanPreferencesKey("onboarding-already-have-an-account")
     val onboardingSplashCarousel = booleanPreferencesKey("onboarding-splash-carousel")
-    val onboardingUseCase = booleanPreferencesKey("onboarding-splash-carousel")
+    val onboardingUseCase = booleanPreferencesKey("onboarding-use-case")
```
Strange to see this change in this PR.

Florian14 replied:
I observed that the key was duplicated so I changed it in a commit, I agree this is not related to this PR but not sure it will make sense to revert the commit? 🤔

bmarty replied:
This is definitely an issue, but maybe (next time) create a quick dedicated PR which can me merged super fast (at least faster than this one). You can even create it from GitHub if you do not want to switch branch locally.

bmarty replied:
(done in #6456)

bmarty reviewed (changes requested):
I focused my review on the SDK part.

bmarty commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/create/CreateLocalRoomTask.kt:
```diff
@@ -0,0 +1,265 @@
+     *
+     * @return a fake event
+     */
+    private fun createFakeEvent(type: String?, content: Content?, stateKey: String? = ""): Event {
```
`Fake` is maybe reserved to test. Rename to `createLocalEvent`?

bmarty commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/create/CreateLocalRoomTask.kt:
```diff
@@ -0,0 +1,265 @@
+                stateKey = stateKey,
+                content = content,
+                originServerTs = clock.epochMillis(),
+                eventId = UUID.randomUUID().toString()
```
`eventId` should start with `!`, maybe better to follow this convention.

bmarty commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/create/CreateLocalRoomTask.kt:
```diff
@@ -0,0 +1,265 @@
+            createLocalRoomSummaryEntity(realm, roomId, createRoomBody)
+        }
+
+        // Wait for room to be created in DB
```
Maybe just replacing with this line can help (I let you do some test):
```koltin
        // Ensure that Realm is up to date before returning the roomId.
        monarchy.doWithRealm { it.refresh() }
```

gaelledel reviewed (approved):
Approved given we change the title to "Direct message" (singular)

Florian14 commented on matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/room/model/localecho/RoomLocalEcho.kt:
```diff
@@ -0,0 +1,28 @@
+    fun isLocalEchoId(roomId: String) = roomId.startsWith(PREFIX)
+
+    fun createLocalEchoId() = "${PREFIX}${UUID.randomUUID()}"
+}
```
I only need the first method public so I changed the second one to internal

bmarty reviewed (approved):
Thanks for the update!
